File: src/cleansia_android/core/src/main/java/cz/cleansia/core/servicearea/ServiceAreaDataSource.kt
package cz.cleansia.core.servicearea

/**
 * App-side adapter contract — each consumer app (customer, partner)
 * implements this by calling its own NSwag-generated `CountryApi.getServiced()`
 * and `ServiceCityApi.getServiceCities(countryId)` and mapping the
 * results to the `:core` slim DTOs. The shared
 * [ServiceAreaProvider] talks only to this interface, so it doesn't
 * care which app it lives inside.
 *
 * Implementations should:
 *  - return NULL on failure (network error, non-2xx) and log it — never throw.
 *    Null is load-bearing: it tells the provider the answer is UNKNOWN, so the
 *    failure is not cached and callers can fail open. Returning an empty list
 *    on failure is indistinguishable from "the company serves nothing" and
 *    used to poison the address picker for the process lifetime.
 *  - lowercase + alpha-2-normalise ISO codes when filling
 *    [ServicedCountry.isoCode] (see IsoCountryCodes) so Mapbox comparisons and
 *    the country= bias param work without call-site tolerance.
 */
interface ServiceAreaDataSource {
    suspend fun fetchServicedCountries(): List<ServicedCountry>?

    /**
     * [countryId] null = "all serviced cities across countries". When
     * provided, the implementation MAY filter server-side or in-memory
     * — the provider caches the unfiltered list either way.
     */
    suspend fun fetchServiceCities(countryId: String?): List<ServicedCity>?
}

File: src/cleansia_ios/CleansiaCore/Sources/CleansiaCore/ServiceArea/ServiceAreaDataSource.swift
import Foundation

/// App-side adapter contract — each app implements this over its own generated
/// serviced-countries call (`Country/GetServiced`) and maps the result to the
/// Core `ServicedCountry` shape. A failure comes back as the ADR-0011
/// `ApiResult` error, which the provider treats as UNKNOWN — never as "serves
/// nothing" — and does not cache.
public protocol ServiceAreaDataSource {
    func fetchServicedCountries() async -> ApiResult<[ServicedCountry]>

    /// Cities the company serves, optionally narrowed to one country. Same UNKNOWN-on-failure contract
    /// as the countries call: a failure is "could not check", never "serves nowhere".
    func fetchServicedCities(countryId: String?) async -> ApiResult<[ServicedCity]>
}

public extension ServiceAreaDataSource {
    /// Defaults to UNKNOWN, so an app that has no reason to ask does not have to answer.
    ///
    /// Only the CUSTOMER books, so only the customer app needs city-level serviceability; the partner
    /// side shows country status and nothing finer. The default is deliberately a failure rather than
    /// an empty list — empty would mean "serves nowhere", which every caller is required to render as a
    /// refusal, and that is the one answer this must never invent.
    func fetchServicedCities(countryId _: String?) async -> ApiResult<[ServicedCity]> {
        .failure(ApiError(code: "servicearea.cities_not_queried"))
    }
}
